File: 3517-smallest-palindromic-rearrangement-i/3517-smallest-palindromic-rearrangement-i.py
class Solution(object):
    def smallestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        count = [0] * 26

        for ch in s:
            count[ord(ch) - ord('a')] += 1

        left = []
        middle = ''

        for i in range(26):
            left.append(chr(i + ord('a')) * (count[i] // 2))

            if count[i] % 2 == 1:
                middle = chr(i + ord('a'))

        left_half = ''.join(left)

        return left_half + middle + left_half[::-1]
        

File: 3517-smallest-palindromic-rearrangement-i/3517-smallest-palindromic-rearrangement-i.cpp
class Solution {
public:
    string smallestPalindrome(string s) {
        vector<int> frequency(26, 0);

        for (char ch : s) {
            frequency[ch - 'a']++;
        }

        string leftHalf = "";
        char middle = '\0';

        for (int i = 0; i < 26; i++) {
            int pairs = frequency[i] / 2;

            while (pairs > 0) {
                leftHalf += char('a' + i);
                pairs--;
            }

            if (frequency[i] % 2 == 1) {
                middle = char('a' + i);
            }
        }

        string rightHalf = leftHalf;
        reverse(rightHalf.begin(), rightHalf.end());

        if (middle != '\0') {
            return leftHalf + middle + rightHalf;
        }

        return leftHalf + rightHalf;
    }
};